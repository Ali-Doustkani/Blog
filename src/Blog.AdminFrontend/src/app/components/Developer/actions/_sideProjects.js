import uuid from 'uuid/v1'

const initSideProjectErrors = sideProject => ({
   ...sideProject,
   titleErrors: [],
   contentErrors: []
})

const newSideProject = () =>
   initSideProjectErrors({
      id: uuid(),
      title: '',
      content: ''
   })

const addSideProject = state => ({
   ...state,
   sideProjects: [...state.sideProjects, newSideProject()]
})

const deleteSideProject = (state, action) => ({
   ...state,
   sideProjects: state.sideProjects.filter(x => x.id !== action.id)
})

const updateSideProject = (state, action) => ({
   ...state,
   sideProjects: state.sideProjects.map(update(action.change))
})

const update = change => sideProject =>
   change.id === sideProject.id ? { ...sideProject, ...change } : sideProject

export { addSideProject, deleteSideProject, updateSideProject, initSideProjectErrors }
