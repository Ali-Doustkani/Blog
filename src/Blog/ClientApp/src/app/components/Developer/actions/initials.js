const STATUS = {
   LOADING: 1,
   SAVING: 2,
   IDLE: 3,
   STOPPED: 4
}

const initialState = {
   status: STATUS.LOADING,
   errorMessage: '',
   summary: '',
   summaryErrors: [],
   skills: '',
   skillsErrors: [],
   experiences: [],
   experiencesErrors: [],
   sideProjects: [],
   sideProjectErrors: []
}

export { initialState, STATUS }
