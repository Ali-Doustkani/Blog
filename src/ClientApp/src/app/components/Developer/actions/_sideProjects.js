import uuid from 'uuid/v1'
import { update } from '../../../utils/fn'

const newSideProject = () => ({
   id: uuid(),
   title: '',
   content: ''
})

const addSideProject = state => ({
   ...state,
   sideProjects: [...state.sideProjects, newSideProject()]
})

const deleteSideProject = (state, action) => ({
   ...state,
   sideProjects: state.sideProjects.filter(x => x.id !== action.id)
})

const updateSideProject = (state, action) => ({
   ...state,
   sideProjects: state.sideProjects.map(update(action.change))
})

export { addSideProject, deleteSideProject, updateSideProject }
