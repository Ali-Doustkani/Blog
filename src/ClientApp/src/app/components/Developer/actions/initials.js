const STATUS = {
   LOADING: 1,
   SAVING: 2,
   IDLE: 3,
   STOPPED: 4
}

const initialState = {
   status: STATUS.LOADING,
   errorMessage: '',
   summary: '',
   skills: '',
   experiences: [],
   sideProjects: [],
   educations: []
}

export { initialState, STATUS }
