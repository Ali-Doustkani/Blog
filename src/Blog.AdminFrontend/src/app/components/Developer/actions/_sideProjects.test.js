import uuid from 'uuid/v1'
import { initialState } from './initials'
import { addSideProject, deleteSideProject, updateSideProject } from './_sideProjects'

jest.mock('uuid/v1')

afterEach(jest.restoreAllMocks)

it('adds a new side project', () => {
   uuid.mockReturnValue(123)
   const initial = {
      sideProjects: []
   }
   const state = addSideProject(initial)

   expect(state.sideProjects).toEqual([{ id: 123, title: '', content: '' }])
})

it('deletes a side project', () => {
   const initial = {
      sideProjects: [
         { id: 1, title: 'Richtext', content: 'Richtext Desc' },
         { id: 2, title: 'CodePrac', content: 'CodePrac Desc' }
      ]
   }

   const state = deleteSideProject(initial, { id: 2 })

   expect(state.sideProjects).toEqual([
      {
         id: 1,
         title: 'Richtext',
         content: 'Richtext Desc'
      }
   ])
})

it('updates a side project', () => {
   const init = {
      sideProjects: [
         { id: 1, title: 'Richtext', content: 'Richtext Desc' },
         { id: 2, title: 'CodePrac', content: 'CodePrac Desc' }
      ]
   }

   let state = updateSideProject(init, { change: { id: 2, title: 'Daily Code Practice' } })
   state = updateSideProject(state, { change: { id: 2, content: 'Description' } })

   expect(state.sideProjects).toEqual([
      { id: 1, title: 'Richtext', content: 'Richtext Desc' },
      { id: 2, title: 'Daily Code Practice', content: 'Description' }
   ])
})

it('keeps the state intact', () => {
   uuid.mockReturnValueOnce(1)
   uuid.mockReturnValueOnce(2)

   let state = addSideProject(initialState)
   state = addSideProject(state)
   state = updateSideProject(state, { change: { id: 1, title: 'Richtext' } })
   state = updateSideProject(state, { change: { id: 1, content: 'Desc' } })
   state = deleteSideProject(state, { id: 2 })

   expect(state).toMatchSnapshot()
})
