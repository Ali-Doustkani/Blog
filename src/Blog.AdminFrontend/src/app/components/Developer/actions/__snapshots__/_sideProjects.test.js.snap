// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`keeps the state intact 1`] = `
Object {
  "errorMessage": "",
  "experiences": Array [],
  "experiencesErrors": Array [],
  "sideProjectErrors": Array [],
  "sideProjects": Array [
    Object {
      "content": "Desc",
      "contentErrors": Array [],
      "id": 1,
      "title": "Richtext",
      "titleErrors": Array [],
    },
  ],
  "skills": "",
  "skillsErrors": Array [],
  "status": 1,
  "summary": "",
  "summaryErrors": Array [],
}
`;
